src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+            i.1.index_op(),
+            i.2.index_op(),
+            i.3.index_op(),
+            i.4.index_op(),
+            i.5.index_op(),
+            i.6.index_op(),
+            i.7.index_op(),
+            i.8.index_op(),
+            i.9.index_op(),
+            i.10.index_op(),
+            i.11.index_op(),
+            i.12.index_op(),
+            i.13.index_op(),
+            i.14.index_op(),
+            i.15.index_op(),
+        ];
+        get_item_nd(self, &i, stream)
+    }
+}
+
+// Implement private bindings
+impl Array {
+    // This is exposed in the c api but not found in the swift or python api
+    //
+    // Thie is not the same as rust slice. Slice in python is more like `StepBy` iterator in rust
+    pub(crate) unsafe fn slice_device_unchecked(
+        &self,
+        start: &[i32],
+        stop: &[i32],
+        strides: &[i32],
+        stream: StreamOrDevice,
+    ) -> Array {
+        unsafe {
+            let c_array = mlx_sys::mlx_slice(
+                self.c_array,
+                start.as_ptr(),
+                start.len(),
+                stop.as_ptr(),
+                stop.len(),
+                strides.as_ptr(),
+                strides.len(),
+                stream.as_ptr(),
+            );
+
+            Array::from_ptr(c_array)
+        }
+    }
+
+    pub(crate) fn try_slice_device(
+        &self,
+        start: &[i32],
+        stop: &[i32],
+        strides: &[i32],
+        stream: StreamOrDevice,
+    ) -> Result<Array, SliceError> {
+        if start.len() != self.ndim() || stop.len() != self.ndim() || strides.len() != self.ndim() {
+            return Err(SliceError { ndim: self.ndim() });
+        }
+
+        unsafe { Ok(self.slice_device_unchecked(start, stop, strides, stream)) }
+    }
+
+    pub(crate) fn slice_device(
+        &self,
+        start: &[i32],
+        stop: &[i32],
+        strides: &[i32],
+        stream: StreamOrDevice,
+    ) -> Array {
+        self.try_slice_device(start, stop, strides, stream).unwrap()
+    }
+}
+
+/* -------------------------------------------------------------------------- */
+/*                              Helper functions                              */
+/* -------------------------------------------------------------------------- */
+
+fn absolute_indices(absolute_start: i32, absolute_end: i32, stride: i32) -> Vec<i32> {
+    let mut indices = Vec::new();
+    let mut i = absolute_start;
+    while (stride > 0 && i < absolute_end) || (stride < 0 && i > absolute_end) {
+        indices.push(i);
+        i += stride;
+    }
+    indices
+}
+
+// Implement additional public APIs
+//
+// TODO: rewrite this in a more rusty way
+#[inline]
+fn gather_nd<'a>(
+    src: &'a Array,
+    operations: impl Iterator<Item = &'a ArrayIndexOp>,
+    gather_first: bool,
+    stream: StreamOrDevice,
+) -> (usize, Array) {
+    use ArrayIndexOp::*;
+
+    let mut max_dims = 0;
+    let mut slice_count = 0;
+    let mut is_slice: Vec<bool> = Vec::new();
```

Comment:
perhaps a small optimization - technically you know the capacity for this

Reply:
Not really right? The `operations` iterator is `Filter` which doesn't know the size beforehand

Reply:
Can't you use `size_hint` here?

Reply:
Well you kind of get the upper bound which would be `last_array_or_index`, but the lower bound would just be 0

Reply:
Pre-allocates using `last_array_or_index` (7cbeb15bd0f7e98efa10df14e3ba47c18521aa32)